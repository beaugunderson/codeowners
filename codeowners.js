// @ts-check
'use strict';

const findUp = require('find-up');
const fs = require('fs');
const ignore = require('ignore');
const isDirectory = require('is-directory');
const path = require('path');
const trueCasePath = require('true-case-path');

function ownerMatcher(pathString) {
  const matcher = ignore().add(pathString);
  return matcher.ignores.bind(matcher);
}

function Codeowners(currentPath) {
  if (!currentPath) {
    currentPath = process.cwd();
  }

  this.codeownersFilePath = trueCasePath(findUp.sync(['.github/CODEOWNERS', 'docs/CODEOWNERS', 'CODEOWNERS'], { cwd: currentPath }));

  this.codeownersDirectory = path.dirname(this.codeownersFilePath);
  // We might have found a bare codeowners file or one inside the two supported subdirectories.
  // In the latter case the project root is up another level.
  if (this.codeownersDirectory.match(/\/(.github|docs)$/i)) {
    this.codeownersDirectory = path.dirname(this.codeownersDirectory);
  }
  const codeownersFile = path.basename(this.codeownersFilePath);

  if (codeownersFile !== 'CODEOWNERS') {
    throw new Error(`Found a CODEOWNERS file but it was lower-cased: ${this.codeownersFilePath}`);
  }

  if (isDirectory.sync(this.codeownersFilePath)) {
    throw new Error(`Found a CODEOWNERS but it's a directory: ${this.codeownersFilePath}`);
  }

  const lines = fs.readFileSync(this.codeownersFilePath).toString().split('\n');
  const ownerEntries = [];

  for (const line of lines) {
    if (!line) {
      continue;
    }

    if (line.startsWith('#')) {
      continue;
    }

    const [pathString, ...usernames] = line.split(/\s+/);

    ownerEntries.push({
      path: pathString,
      usernames: usernames,
      match: ownerMatcher(pathString)
    });
  }

  this.ownerEntries = ownerEntries;
}

Codeowners.prototype.getOwner = function (filePath) {
  let owners = [];
  for (const entry of this.ownerEntries) {
    if (entry.match(filePath)) {
      owners = entry.usernames;
    }
  }
  return owners.slice();
}

module.exports = Codeowners;
