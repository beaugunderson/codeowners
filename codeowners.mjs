import findUp from "find-up";
import fs from "fs";
import path from "path";
import ignore from "ignore";
import isDirectory from "is-directory";

const SECTION_REGEX = /\[([^\]]*)\](?:\[(\d+)\])?([^#]*)(?:#(.*))?/;
const ENTRY_REGEX = /^((?:\\\s|[^\s])*)\s?([^#]*)(#.*)?$/;

/**
 * @param {string} pathString the path to match
 * @returns {function(): boolean}
 */
function ownerMatcher(pathString) {
  const matcher = ignore().add(pathString);
  return matcher.ignores.bind(matcher);
}

export default function Codeowners(currentPath, fileName = "CODEOWNERS") {
  const pathOrCwd = currentPath || process.cwd();

  this.codeownersFilePath = findUp.sync(
    [
      `.github/${fileName}`,
      `.gitlab/${fileName}`,
      `docs/${fileName}`,
      `${fileName}`,
    ],
    { cwd: pathOrCwd }
  );

  if (!this.codeownersFilePath) {
    throw new Error(`Could not find a CODEOWNERS file`);
  }

  this.codeownersDirectory = path.dirname(this.codeownersFilePath);

  // We might have found a bare codeowners file or one inside the three supported subdirectories.
  // In the latter case the project root is up another level.
  if (this.codeownersDirectory.match(/\/(.github|.gitlab|docs)$/i)) {
    this.codeownersDirectory = path.dirname(this.codeownersDirectory);
  }

  const codeownersFile = path.basename(this.codeownersFilePath);

  if (codeownersFile !== fileName) {
    throw new Error(
      `Found a ${fileName} file but it was lower-cased: ${this.codeownersFilePath}`
    );
  }

  if (isDirectory.sync(this.codeownersFilePath)) {
    throw new Error(
      `Found a ${fileName} but it's a directory: ${this.codeownersFilePath}`
    );
  }

  const lines = fs
    .readFileSync(this.codeownersFilePath)
    .toString()
    .split(/\r\n|\r|\n/)
    .map((line) => line.trim())
    .filter((line) => line.length && !line.startsWith("#"));

  const ownerSections = [];
  let currentSection = { name: null, owners: [], entries: [] };

  for (const line of lines) {
    if (SECTION_REGEX.test(line)) {
      const groups = SECTION_REGEX.exec(line);
      const name = groups[1];
      const owners = (groups[3] || "").split(/\s+/).filter((v) => v.length);

      ownerSections.push(currentSection);
      currentSection = { name, owners, entries: [] };
      continue;
    }
    if (ENTRY_REGEX.test(line)) {
      const groups = ENTRY_REGEX.exec(line);
      const pathString = groups[1].replace(/\\\s/g, " ");
      const usernames = (groups[2] || "").split(/\s+/).filter((v) => v.length);

      currentSection.entries.push({
        path: pathString,
        usernames: usernames.length
          ? [...usernames]
          : [...currentSection.owners],
        match: ownerMatcher(pathString),
      });
      continue;
    }
    console.error(`Could not parse line: ${line}`);
  }

  ownerSections.push(currentSection);

  // reverse the owner entries to search from bottom to top
  // the last matching pattern takes the most precedence
  this.ownerSections = ownerSections.reverse();
}

Codeowners.prototype.getOwner = function getOwner(filePath) {
  for (const section of this.ownerSections) {
    const owners = [];
    for (const entry of section.entries) {
      if (entry.match(filePath)) {
        owners.push(...entry.usernames);
      }
    }
    if (owners.length) {
      return [...new Set(owners)];
    }
  }

  return [];
};
