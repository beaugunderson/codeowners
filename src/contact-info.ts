/**
 * Class for indexing and accessing contact info listed in a codeowners file.
 */
export class ContactInfo {
  fields: string[] = [];
  owners: Array<Record<string, string>> = [];

  /**
   * Consume one line of contact info
   * @param line Contact info line, with leading '##' intact
   */
  addLine(line: string) {
    const dataItems = line.replace(/^##\s*/, "").split(" ");
    if (!this.fields.length) {
      // assume first line defines column field names
      this.fields = dataItems;
      return;
    }
    const ownerInfo: Record<string, string> = {};
    for (const field of this.fields) {
      const value = dataItems.shift();
      if (value) {
        ownerInfo[field] = value;
      }
    }
    this.owners.push(ownerInfo);
  }
}
