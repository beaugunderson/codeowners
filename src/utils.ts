import { statSync } from "node:fs";

export function isDirectorySync(filepath: string) {
  if (typeof filepath !== "string") {
    throw new Error("expected filepath to be a string");
  }
  const stat = statSync(filepath, { throwIfNoEntry: false });
  if (!stat) return false;
  return stat.isDirectory();
}

export function times<T>(count: number, cb: (n: number) => T, indexAt = 0): Array<T> {
  const ret: T[] = [];
  for (let i = indexAt; i < count + indexAt; i++) {
    ret.push(cb(i));
  }
  return ret;
}

export function padEnd(input: string, length: number): string {
  const strLength = length ? input.length : 0;
  return length && strLength < length ? input + times(length - strLength, () => " ") : input || "";
}

export function intersection<T>(a: Array<T>, b: Array<T>): Array<T> {
  const aItems = new Set(a);
  const ret: Array<T> = [];
  for (const item of b) {
    if (aItems.has(item)) {
      ret.push(item);
    }
  }
  return ret;
}
