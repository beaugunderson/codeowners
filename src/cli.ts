#!/usr/bin/env node

import fs from "node:fs";
import path from "node:path";
import { walkStream } from "@nodelib/fs.walk";
import { program } from "commander";
import { findUpSync } from "find-up";
import ignore from "ignore";

import { Codeowners } from "./codeowners";
import { intersection, padEnd } from "./utils";

const rootPath = process.cwd();

const gitignorePath = findUpSync(".gitignore", { cwd: rootPath });
const gitignoreMatcher = ignore();

if (gitignorePath) {
  gitignoreMatcher.add(fs.readFileSync(gitignorePath).toString());
}

program.description("@nmann/codeowners cli");

program
  .command("of <file> [otherFiles...]")
  .description("list the owner of specific files")
  .action((singleFile, otherFiles) => {
    const codeowners = new Codeowners(rootPath);
    console.log(singleFile, ...codeowners.getOwner(singleFile));
    for (const file of otherFiles) {
      console.log(file, ...codeowners.getOwner(file));
    }
  });

program
  .command("audit")
  .description("list the owners for all files")
  .option("-u, --unowned", "unowned files only")
  .option(
    "-d, --min-depth <n>",
    "roll up unowned files to common paths, but to a minimum depth (only works for unowned)",
    parseInt,
  )
  .option("-w, --width <columns>", "how much should filenames be padded?", "32")
  .option(
    "-c, --codeowners-filename <codeowners_filename>",
    "specify CODEOWNERS filename",
    "CODEOWNERS",
  )
  .action((options) => {
    let codeowners: Codeowners;

    try {
      codeowners = new Codeowners(rootPath, options.codeownersFilename);
    } catch (e) {
      console.error((e as Error).message);
      process.exit(1);
    }

    const padding = parseInt(options.width, 10);

    const stream = walkStream(rootPath, {
      deepFilter: (entry) => {
        const split = entry.path.split(path.sep);
        return (
          !split.includes("node_modules") && !split.includes(".git") && !split.includes(".cache")
        );
      },
      errorFilter: (error) =>
        error.code === "ENOENT" || error.code === "EACCES" || error.code === "EPERM",
    });

    stream.on("data", (file) => {
      let relative = path
        .relative(codeowners.codeownersDirectory, file.path)
        .replace(/(\r)/g, "\\r");
      if (gitignoreMatcher.ignores(relative)) {
        return;
      }

      let owners = codeowners.getOwner(relative);

      if (options.unowned) {
        if (!owners.length) {
          if (options.minDepth) {
            while (!owners.length && relative.split(path.sep).length > options.minDepth) {
              const parentPath = path.dirname(relative);
              owners = codeowners.getOwner(parentPath);
              if (!owners.length) {
                relative = parentPath;
              } else {
                break;
              }
            }
          }

          console.log(relative);
        }
      } else {
        let printedOwner = "nobody";
        if (owners.length) {
          printedOwner = owners.join(" ");
        }
        console.log(`${padEnd(relative, padding)}    ${printedOwner}`);
      }
    });

    stream.on("error", (err) => {
      console.error(err);
    });
  });

program
  .command("verify <path> <users...>")
  .description("verify users/teams own a specific path")
  .option(
    "-c, --codeowners-filename <codeowners_filename>",
    "specify CODEOWNERS filename",
    "CODEOWNERS",
  )
  .action((checkPath, users, options) => {
    let codeowners: Codeowners;

    // instantiate new Codeowners obj
    try {
      codeowners = new Codeowners(rootPath, options.codeownersFilename);
    } catch (e) {
      console.error((e as Error).message);
      process.exit(1);
    }

    // call getOwner() on `path`
    const owners = codeowners.getOwner(checkPath);

    // check if any `users` are in the results of getOwner()
    const verifiedOwners = intersection(users, owners);

    // if verifiedOwners is empty, exit with error
    if (verifiedOwners.length < 1) {
      console.log(`None of the users/teams specified own the path ${checkPath}`);
      process.exit(1);
    }

    // print owners
    for (const currOwner of verifiedOwners) {
      console.log(`${checkPath}    ${currOwner}`);
    }
  });

if (!process.argv.slice(2).length) {
  program.outputHelp();
}

program.parse(process.argv);
