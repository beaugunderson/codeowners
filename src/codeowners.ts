import fs from "node:fs";
import path from "node:path";
import { findUpSync } from "find-up";
import ignore from "ignore";
import * as tcp from "true-case-path";

import { ContactInfo } from "./contact-info";
import { isDirectorySync } from "./utils";

function ownerMatcher(pathString: string) {
  const matcher = ignore().add(pathString);
  return matcher.ignores.bind(matcher);
}

const PARENT_FOLDERS = [".github", ".gitlab", "docs"];
const CODEOWNERS = "CODEOWNERS";

interface OwnerEntry {
  path: string;
  usernames: string[];
  match(pathname: string): boolean;
}

interface ReadOnlyDict<T> {
  readonly [key: string]: T | undefined;
}

/**
 * @param currentPath defaults to process.cwd()
 * @param fileName of file to find, defaults to CODEOWNERS
 */
export class Codeowners {
  readonly codeownersFilePath: string;
  readonly codeownersDirectory: string;
  readonly contactInfo: ReadonlyArray<ReadOnlyDict<string>>;

  private ownerEntries: OwnerEntry[] = [];
  private pathsByOwner: Record<string, string[]> = {};

  constructor(currentPath: string = process.cwd(), fileName: string = CODEOWNERS) {
    const contactInfo = new ContactInfo();

    const foundPath = findUpSync(
      PARENT_FOLDERS.map((folder) => path.join(folder, fileName)).concat(fileName),
      {
        cwd: currentPath,
      },
    );

    if (!foundPath) {
      throw new Error("Could not find a CODEOWNERS file");
    }

    this.codeownersFilePath = tcp.trueCasePathSync(foundPath);
    this.codeownersDirectory = path.dirname(this.codeownersFilePath);

    // We might have found a bare codeowners file or one inside the three supported subdirectories.
    // In the latter case the project root is up another level.
    if (PARENT_FOLDERS.includes(path.basename(this.codeownersDirectory))) {
      this.codeownersDirectory = path.dirname(this.codeownersDirectory);
    }

    const codeownersFile = path.basename(this.codeownersFilePath);

    if (codeownersFile !== fileName) {
      throw new Error(
        `Found a ${fileName} file but it was lower-cased: ${this.codeownersFilePath}`,
      );
    }

    if (isDirectorySync(this.codeownersFilePath)) {
      throw new Error(`Found a ${fileName} but it's a directory: ${this.codeownersFilePath}`);
    }

    const lines = fs
      .readFileSync(this.codeownersFilePath)
      .toString()
      .split(/\r\n|\r|\n/);

    for (const line of lines) {
      if (!line) {
        continue;
      }

      if (line.startsWith("##")) {
        contactInfo.addLine(line);
        continue;
      }

      if (line.startsWith("#")) {
        // ignore comment
        continue;
      }

      const [pathString, ...usernames] = line.split(/\s+/);
      const matcher = ownerMatcher(pathString);
      this.ownerEntries.push({
        path: pathString,
        usernames,
        match(pathname) {
          return matcher(path.relative(currentPath, pathname));
        },
      });
      for (const owner of usernames) {
        if (!this.pathsByOwner[owner]) {
          this.pathsByOwner[owner] = [];
        }
        this.pathsByOwner[owner].push(pathString);
      }
    }

    // reverse the owner entries to search from bottom to top
    // the last matching pattern takes the most precedence
    this.ownerEntries.reverse();
    this.contactInfo = contactInfo.owners;
  }

  getOwner(filePath: string) {
    for (const entry of this.ownerEntries) {
      if (entry.match(filePath)) {
        return [...entry.usernames];
      }
    }

    return EMPTY_ARRAY;
  }

  getPathsForOwner(owner: string) {
    if (this.pathsByOwner[owner]) {
      return this.pathsByOwner[owner].slice();
    }
    return [];
  }
}

export default Codeowners;

const EMPTY_ARRAY: never[] = [];
