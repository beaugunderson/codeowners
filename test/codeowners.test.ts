import { describe, expect, it } from "bun:test";

import { Codeowners } from "../src/codeowners";

const repos = new Codeowners();

describe("codeowners", () => {
  it("returns owners for file", () => {
    const owner = repos.getOwner(__filename);
    expect(owner).toEqual(["@noahm"]);
  });

  it("owners is a copy of internal data", () => {
    repos.getOwner(__filename).pop();

    const owner = repos.getOwner(__filename);
    expect(owner).toEqual(["@noahm"]);
  });
});

describe("metadata", () => {
  it("parses metadata", () => {
    const metadata = repos.contactInfo;
    expect(metadata).toMatchSnapshot();
  });
});
