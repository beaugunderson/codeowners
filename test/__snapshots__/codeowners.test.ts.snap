// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`metadata parses metadata 1`] = `
[
  {
    "ice-cream-flavor": "strawberry",
    "owner": "@noahm",
    "timezone": "PST",
  },
  {
    "ice-cream-flavor": "chocolate",
    "owner": "@amoose136",
    "timezone": "EST",
  },
]
`;
